### PR TITLE
Prefer ldlogger over intercept-build

### LINKED ISSUE #1091: ldlogger instead of intercept build
intercept-build should be used only as a fallback if ldlogger is not available in `CodeChecker log`

at package build time if `intercept-build` is available in the path do not build ldlogger we could add an extra argument to build the ldlogger even if the intercept-build is in the path

### PR DESCRIPTION
**Pull request by steakhal**
> Closes #1091

1st check if LDlogger is available and use that.
2nd check if intercept-build is available and use that.
3rd Otherwise report an error.

### CODE REVIEW

File: analyzer/tests/functional/analyze_and_parse/__init__.py [hunk 1]
```diff
@@ -26,6 +26,11 @@ def setup_package():
 
     os.environ['TEST_WORKSPACE'] = TEST_WORKSPACE
 
+    # TODO: We should somehow define the 'CC_LIB_DIR' envvar, to
+    # let the host_check.check_ldlogger() detect the ldlogger.
+    # Alternatively, we could use the 'CC_DATA_FILES_DIR' envvar as well.
+    os.environ['CC_LIB_DIR'] = "TODO"
+
```
Comment (csordasmarton) [inline analyzer/tests/functional/analyze_and_parse/__init__.py:33]: This is not needed, the CC_LIB_DIR is set by the top level script file:
```suggestion
```

File: analyzer/codechecker_analyzer/buildlog/host_check.py [hunk 2]
```diff
@@ -48,3 +50,30 @@ def check_intercept(env):
...
+    lib_dir_path = env.get('CC_LIB_DIR', '')
+
+    if not data_files_dir_path and not lib_dir_path:
+        LOG.debug('Both CC_DATA_FILES_DIR and CC_LIB_DIR environment '
+                  'variables as undefined or empty in '
+                  'the logger environment.')
+        return False
+
+    logger_lib_dir_path = os.path.join(
+            data_files_dir_path, 'ld_logger', 'lib', '*', 'ldlogger.so')
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/host_check.py:70]:
```suggestion
    logger_lib_dir_path = os.path.join(
        data_files_dir_path, 'ld_logger', 'lib', '*', 'ldlogger.so')
```

File: analyzer/codechecker_analyzer/buildlog/host_check.py [hunk 3]
```diff
@@ -48,3 +50,30 @@ def check_intercept(env):
             LOG.debug(oerr)
             LOG.debug('Failed to run: "%s"', ' '.join(intercept_cmd))
             return False
+
+
+def check_ldlogger(env):
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/host_check.py:55]:
```suggestion
def check_ldlogger(env) -> bool:
```

File: analyzer/codechecker_analyzer/buildlog/build_manager.py [hunk 4]
```diff
@@ -76,18 +76,8 @@ def perform_build_command(logfile, command, context, keep_link, silent=False,
...
-        command = ' '.join(["intercept-build",
-                            "--cdb", logfile,
-                            "sh -c", shlex.quote(final_command)])
-        log_env = original_env
-        LOG.debug_analyzer(command)
-
-    # Run user's commands in shell.
-    else:
+    # Run user's commands in shell, and preload ldlogger.
+    if host_check.check_ldlogger(original_env):
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/build_manager.py:80]: The original_env doesn't contain environment variable which are set by the top level script and using this function. For this reason use the current environment:
```suggestion
    if host_check.check_ldlogger(os.environ):
```

File: analyzer/codechecker_analyzer/buildlog/build_manager.py [hunk 5]
```diff
@@ -123,6 +115,19 @@ def perform_build_command(logfile, command, context, keep_link, silent=False,
             LOG.error("Intercept-build is required to run CodeChecker in "
                       "OS X.")
             sys.exit(1)
+    # Otherwise try to fall back to intercept-build.
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/build_manager.py:118]: Now the logger detection looks like this:
```py
if ld_logger:
  if LINUX:
    ...
  elif WINDOWS:
    ...
  else:
    ...
elif intercept_build:
  ...
else:
  ...
```

Wouldn't it be better to use the following check:
```py
if ld_logger:
  ...
elif intercept_build:
  ...
else: # No intercept build or ld-logger
  if LINUX:
    ...  # Install ld-logger or intercept-build
  elif WINDOWS:
    ...  # Original log message.
  else:
    ...  # Install intercept-build
```

**Review by csordasmarton — CHANGES_REQUESTED** [review]

**Comment by steakhal** [comment]
I don't want to make further changes, since the CI is already breaks.

File: analyzer/codechecker_analyzer/buildlog/host_check.py [hunk 6]
```diff
@@ -48,3 +50,77 @@ def check_intercept(env):
...
+        data_files_dir_path, 'ld_logger', 'lib', '*', 'ldlogger.so')
+
+    LOG.debug(f'logger_lib_dir_path: {logger_lib_dir_path}')
+    asd = glob.glob(logger_lib_dir_path)
+    LOG.debug(f'glob.glob(logger_lib_dir_path): {asd}')
+    if asd:
+        return True
+
+    alternative_logger_lib_dir_path = os.path.join(
+        lib_dir_path, 'codechecker_analyzer', 'ld_logger', 'lib', '*',
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/host_check.py:119]:
```suggestion
        lib_dir_path, 'codechecker_analyzer', 'ld_logger', 'lib', '**',
```

File: analyzer/codechecker_analyzer/buildlog/host_check.py [hunk 7]
```diff
@@ -48,3 +50,77 @@ def check_intercept(env):
...
+        'ldlogger.so')
+
+    LOG.debug(
+        f'alternative_logger_lib_dir_path: {alternative_logger_lib_dir_path}')
+    asd = glob.glob(alternative_logger_lib_dir_path)
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/host_check.py:124]:
```suggestion
    asd = glob.glob(alternative_logger_lib_dir_path, recursive=True)
```

File: analyzer/codechecker_analyzer/buildlog/host_check.py [hunk 8]
```diff
@@ -48,3 +50,77 @@ def check_intercept(env):
...
+        pass
+    try:
+        txt = os.listdir(os.path.join(lib_dir_path, 'codechecker_analyzer',
+                                      'ld_logger', 'lib'))
+        LOG.debug(f'listdir 2-4: {txt}')
+    except:
+        pass
+
+    logger_lib_dir_path = os.path.join(
+        data_files_dir_path, 'ld_logger', 'lib', '*', 'ldlogger.so')
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/host_check.py:110]:
```suggestion
        data_files_dir_path, 'ld_logger', 'lib', '**', 'ldlogger.so')
```

File: analyzer/codechecker_analyzer/buildlog/host_check.py [hunk 9]
```diff
@@ -48,3 +50,77 @@ def check_intercept(env):
...
+
+    LOG.debug(f'logger_lib_dir_path: {logger_lib_dir_path}')
+    asd = glob.glob(logger_lib_dir_path)
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/buildlog/host_check.py:113]:
```suggestion
    asd = glob.glob(logger_lib_dir_path, recursive=True)
```

**Review by csordasmarton — APPROVED** [review]
LGTM!

**Pull request merged**